### DISCUSSION TITLE
Is it possible to clone a specific folder of a Repository using git

### DISCUSSION THREAD

**Discussion by ghost** [community/community#102639] [Programming Help] [answered]
### Body

# Help Needed
Hey fellows savy devs. Just want to know the following.

*HOW TO CLONE A SPECIFIC DIRECTORY OF A REPOSITORY USING GIT*

Searched for tuts but nothing helped me.

## Why?
I have created a [flask bolerplate](https://github.com/Yuheshpandian/Flask-Boilerplate) to fasten up flask project development. So I want to include all the boilerplate code in a single directory so that i could avoid users cloning unwanted files like README.md, Contributing.md, etc.

`Less Files Less Space`

## What you could do
If this is not possible can soeone state what i could do instead for this action. I thought of REALISING versions of the boilerplate in Releases in Github...

Hope someone would help : )

CHECKED FOR EXISTING DICUSSIONS BUT NOTHING RELAVENT FOUND

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by ghost** [comment]
Please provide step by step instructions. As Im not that wise...  : )

**Reply by pakeku** [reply]
### ✅ Steps

1.  **Clone the repository (without checking out files yet)**

    `git clone --no-checkout https://github.com/shadcn-ui/ui.git
    cd ui`

2.  **Enable sparse checkout in cone mode**

    `git sparse-checkout init --cone`

3.  **Specify the folder you want**

    `git sparse-checkout set apps/www`

4.  **Check out the branch**

    `git checkout main`

* * * * *

### 📂 Verify the result

List your files to confirm only the desired folder is present:

`ls`

You should see:

`apps/`

Then check inside:

`ls apps`

Output:

`www/`

* * * * *

### 🧹 Optional cleanup

To remove unnecessary cached files and reduce space:

`git clean -fdx
git gc --prune=now`

* * * * *

✅ **Result:**\
You now have only the `apps/www` folder from the repository --- no extra files, no full repo download.

**Comment by AleixMT** [comment, marked as answer]
Hi there [Yuheshpandian](https://github.com/Yuheshpandian). Yes, you can clone a repository and only checkout (download) certain folders. This operation is called Sparse Checkout and was introduced with Git 1.7.0

The full instructions are found [here](https://stackoverflow.com/questions/600079/is-there-any-way-to-clone-a-git-repositorys-sub-directory-only).

```shell
git init <repo>
cd <repo>
git remote add -f origin <url>

git config core.sparseCheckout true

# This tells git which directories you want to checkout. Then you can pull just those directories
echo "some/dir/" >> .git/info/sparse-checkout
echo "another/sub/tree" >> .git/info/sparse-checkout  

git pull origin master
```

Where `/some/dir` and `another/sub/tree` are the folders that you specifically want to check out. 

There are two alternatives that may also be useful to you:

## Use releases tab of your repo

If your goal is to reduce the size of the download of your repository I think that using the releases tab of your repository is a very good idea. With this strategy users will only download the files that you specifically store in the releases tab. 

The downsides are that you will need to explicitly produce a release every time you would like to update those files. This downside can be comletely avoided using a [GitHub Action to build a release on every commit to master branch](https://stackoverflow.com/questions/64708371/how-to-run-github-workflow-on-every-commit-of-a-push), but this is far more than you asked for. 


## Use a service such as DownGit to download a folder from a repository

Another strategy that you can use is using a service such as [DownGit](https://downgit.evecalm.com/). DownGit is a service that allows you to download **folders** from git repositories and also to generate links that allow you to download a **folder** from a git repository. 

I will explain how to use it step by step as you requested and I will use your repo [Flask Boilerplate](https://github.com/Yuheshpandian/Flask-Boilerplate) as example. In particular, I will be generating a link to download the `static` folder of this repo that you provided as example.

1. First of all navigate to the GitHub page of your repository with your browser:

![image](https://github.com/community/community/assets/23342150/53ad4790-6e0f-4cd4-83b5-f4957562959f)

2. Navigate the files until you are in the folder that you want to download. In my case "static":

![image](https://github.com/community/community/assets/23342150/c65cdc6d-a811-4821-8072-123acbf1d276)

3. Copy the URL of your browser when you are in the folder that you want to download

As you can see in the previous screenshot, the URL is [https://github.com/Yuheshpandian/Flask-Boilerplate/tree/main/static](https://github.com/Yuheshpandian/Flask-Boilerplate/tree/main/static)

4. Then, you need to navigate with your browser to [https://downgit.evecalm.com/](https://downgit.evecalm.com/) and you will see this form:

![image](https://github.com/community/community/assets/23342150/7961e7fc-02dc-4c33-a7a0-c0c0970798ff)

5. Paste the link and choose between `Download` and `Create Download Link` depending on if you want to download the folder now or you prefer to generate a link that you can use later to download the folder. The later option probably is what you are interested in since you can generate a link and put it on your `README.md` so you users can easily download the boilerplate code of your project if you stored it in a folder of the repository. 

![image](https://github.com/community/community/assets/23342150/ca5fb301-47fa-45a0-97b7-4d492a1ed744)


I hope that this answer helps you! 

Please mark my answer as valid if this is useful to you 🙏🏻  Thank you

**Reply by ghost** [reply]
That was really awesome I guess I will go with releases as you suggested but also provide a alternative way by using the git or downgit in the README. You were really helpful thank you for guiding me step by step. Happy Programming Journey.


Once again Thanks ☺️

**Reply by ghost** [reply]
And  thanks for taking time to explain step by step to me.

: )

**Reply by wlamagna** [reply]
Thanks, i was looking for this solution too.

**Reply by gatere-kinyanjui** [reply]
Lovely!

**Comment by ridafkih** [comment]
If you're in the future, and looking for a solution because you're—like me—trying to clone a large repository on a slow connection, on a hotspot where you don't have unlimited, or both, this is what worked best for me.

In my case, I was trying to work on [raycast/extensions](https://github.com/raycast/extensions), which is a massive repository. Here are the commands I used in order not to have to download 10GiB+ of repository where I only wanted to modify a few files at most.

```sh
git clone --filter=blob:none --sparse git@github.com:ridafkih/extensions.git
cd extensions
git sparse-checkout set extensions/unicode-symbols
cd extensions/unicode-symbols
```

Just like that, you should have only have to have downloaded the commit objects, tree objects, but no blobs outside of your targeted folder. 🫡

**Reply by davidchi31415** [reply]
This is the only one that helped me. Thanks.
